Sources/AsyncKit/ConnectionPool.swift

```diff
@@ -138,43 +161,60 @@ public final class ConnectionPool<Source> where Source: ConnectionPoolSource  {
     ///
     /// The connection returned by this method should be released when you are finished using it.
     ///
-    ///     let conn = try pool.requestConnection().wait()
+    ///     let conn = try pool.requestConnection(on: ...).wait()
     ///     defer { pool.releaseConnection(conn) }
     ///     // use the connection
     ///
     /// See `withConnection(_:)` for a callback-based method that automatically releases the connection.
     ///
+    /// - parameters:
+    ///     - on: Preferred event loop for the new connection.
     /// - returns: A future containing the requested connection.
-    public func requestConnection() -> EventLoopFuture<Source.Connection> {
-        guard !self.isClosed else {
-            return self.source.eventLoop.makeFailedFuture(ConnectionPoolError.closed)
+    public func requestConnection(
+        on eventLoop: ConnectionPoolEventLoop = .any
+    ) -> EventLoopFuture<Source.Connection> {
+        // synchronize access to available / active connection checks
+        self.lock.lock()
+        defer { self.lock.unlock() }
+
+        let eventLoop = eventLoop.on(self.eventLoopGroup)
+        guard !self.didShutdown else {
+            return eventLoop.makeFailedFuture(ConnectionPoolError.shutdown)
         }
-        
-        if let conn = self.available.popLast() {
+
+        if let conn = self.available.popFirst() {
             // check if it is still open
             if !conn.isClosed {
                 // connection is still open, we can return it directly
-                return self.source.eventLoop.makeSucceededFuture(conn)
+                return eventLoop.makeSucceededFuture(conn)
             } else {
                 // connection is closed, we need to replace it
-                return self.source.makeConnection().flatMapErrorThrowing { error in
+                return self.source.makeConnection(on: eventLoop).flatMapErrorThrowing { error in
```

Comment:
This branch is almost identical to the `if self.activeConnections < self.configuration.maxConnections` one below; could the two be somehow be merged?